### ISSUE TITLE
iocage [upgrade|fetch] does not take proxy settings into account

### ISSUE THREAD

**Issue by ableischwitz** [iocage/iocage#1159]
Make sure to follow and check these boxes before submitting an issue! Thank you.

- [X] Supply `iocage --version`
```
# iocage --version
Version	FreeNAS 11.2-U6
```

- [X] Supply the commands used, along with any steps to recreate it.
- [X] Provide the output from the command you issued.
Run
```
# setenv | grep -i proxy
HTTPS_PROXY=http://<host>:<port>
HTTP_PROXY=http://<host>:<port>
### the following ones may be not relevant
http_proxy=http://<host>:<port>
https_proxy=http://<host>:<port>

# iocage upgrade <jail_with_older_release> -r 11.2-RELEASE
Looking up update.FreeBSD.org mirrors... 3 mirrors found.
Fetching metadata signature for 11.1-RELEASE from update1.freebsd.org... done.
Fetching metadata index... done.
Fetching 1 metadata patches. done.
Applying metadata patches... done.
Fetching 1 metadata files... done.
Inspecting system... done.

The following components of FreeBSD seem to be installed:
src/src world/base world/doc world/lib32

The following components of FreeBSD do not seem to be installed:
world/base-dbg world/lib32-dbg


Fetching metadata signature for 11.2-RELEASE from update1.freebsd.org... done.
Fetching metadata index... done.
Fetching 1 metadata patches. done.
Applying metadata patches... done.
Fetching 1 metadata files... done.
Inspecting system... done.
Fetching files from 11.1-RELEASE for merging... done.
Preparing to download files... done.
Fetching 42999 patches.....10....20....30....40....50....60 done.
Applying patches... done.
Fetching 46594 files... ....10... failed.
Traceback (most recent call last):
  File "/usr/local/bin/iocage", line 10, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/iocage_cli/upgrade.py", line 39, in cli
    ioc.IOCage(jail=jail, skip_jails=True).upgrade(release)
  File "/usr/local/lib/python3.6/site-packages/iocage_lib/iocage.py", line 1805, in upgrade
    root_path).upgrade_jail()
  File "/usr/local/lib/python3.6/site-packages/iocage_lib/ioc_upgrade.py", line 116, in upgrade_jail
    stdin=su.PIPE, env=self.upgrade_env)
RuntimeError: Error occured, jail not upgraded!
```
When inspecting the proxy-logs, no request could be seen from that activitiy.

Also fetching may fail:
```
# iocage fetch -r 11.2-RELEASE
Fetching: 11.2-RELEASE

Extracting: base.txz... 
Extracting: lib32.txz... 
Extracting: doc.txz... 
Extracting: src.txz... 

* Updating 11.2-RELEASE to the latest patch level... 
Looking up update.FreeBSD.org mirrors... 3 mirrors found.
Fetching metadata signature for 11.2-RELEASE from update2.freebsd.org... failed.
Fetching metadata signature for 11.2-RELEASE from update1.freebsd.org... failed.
Fetching metadata signature for 11.2-RELEASE from update4.freebsd.org... done.
Fetching metadata index... done.
Inspecting system... done.
Preparing to download files... done.
Fetching 773 patches.....10....20. done.
Applying patches... done.
Fetching 852 files...  failed.
No updates are available to install.
Run '/tmp/tmp3ecrkdus fetch' first.
```
Again no entries in the proxy-logs.

- [X] Supply what you expected the result or output to be
`iocage [fetch|upgrade]` should take `HTTP_PROXY*` settings into account and use them in the environment of `Popen` when calling sub-commands.

- [X] Checked that the problem has not already been fixed on `master` if using
 a stable release.
The current  source code still does not provide those settings for the sub-commands invoked

**Comment by yo000** [comment]
Please reopen, as the problem still occurs with "iocage upgrade myjail -r 12.2-RELEASE" for example.
The issue seems to be in IOCExec init where proxy vars are not imported.

Please see a patch proposal attached

[iocage_upgrade_behind_proxy.patch.txt](https://github.com/iocage/iocage/files/6034928/iocage_upgrade_behind_proxy.patch.txt)

**Comment by nxg** [comment]
I notice that the associated patch respects environment variable `$HTTP_PROXY`.  It would be be useful for it to instead (or, if necessary, additionally) respect `$http_proxy` (ie, lowercase).

There is no Standard for this, and it is indeed relatively unusual for an exported environment variable to be lowercase, but the lowercase version does seem to be more conventional, and is more broadly supported.  And, for what it's worth, it has been conventional for a couple of decades.

There is some discussion of this on [a Stackoverflow question](https://unix.stackexchange.com/questions/212894/whats-the-right-format-for-the-http-proxy-environment-variable-caps-or-no-ca), which is useful in part because it points to other posts ([why `curl` switched to using _exclusively_ lowercase](https://everything.curl.dev/usingcurl/proxies/env), and [a commendably systematic survey of what's supported where](https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/)).

**Comment by nxg** [comment]
Indeed, can I suggest the attached patch: [ioc_exec.txt](https://github.com/iocage/iocage/files/7936173/ioc_exec.txt)?  This is as applied to `iocage_lib/ioc_exec.py` in released version 1.2 (FreeBSD 13.0).

This defines `{http,https,no}_proxy` in the IOCExec environment, taking the value from the program environment (as with the other patch proposed in this issue), but examining both the lowercased and uppercased variables, and preferring the lowercased variable if both are present.

**Comment by ableischwitz** [comment]
> I notice that the associated patch respects environment variable `$HTTP_PROXY`. It would be be useful for it to instead (or, if necessary, additionally) respect `$http_proxy` (ie, lowercase).
> 
This issue is just addressing the fetch/update process of iocage. Both use methods which only take the uppercased environment variables into account. So technically no need to export more than required.

> There is no Standard for this, and it is indeed relatively unusual for an exported environment variable to be lowercase, but the lowercase version does seem to be more conventional, and is more broadly supported. And, for what it's worth, it has been conventional for a couple of decades.
> 
> There is some discussion of this on [a Stackoverflow question](https://unix.stackexchange.com/questions/212894/whats-the-right-format-for-the-http-proxy-environment-variable-caps-or-no-ca), which is useful in part because it points to other posts ([why `curl` switched to using _exclusively_ lowercase](https://everything.curl.dev/usingcurl/proxies/env), and [a commendably systematic survey of what's supported where](https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/)).

As long as the used methods rely on the *_PROXY variables, there is no need to expose those to the sub-process. Once the methods would change on FreeBSD level, there would be adjustments needed here anyways.

**Comment by nxg** [comment]
> This issue is just addressing the fetch/update process of iocage.
> Both use methods which only take the uppercased environment variables into account. So technically no need to export more than required.

I'm puzzled.  The fetch process didn't work for me with the existing code, with `$http_proxy` defined in the environment, but it did work when I hacked in the attached patch (which, note, defines only the lowercased variables in `su_env`, instead of the uppercased ones, unlike the other patch in the issue, which defines only the uppercased ones).  So _something_ under there did seem to be respecting the lowercased proxy variables and not the uppercased ones.  Does the process run `curl` under the hood, perhaps?

It's quite possible, of course, that there's some version skew, since I was using released v1.2 rather than a repo version; or that I'm potentially getting confused between variables in the (outermost) environment and those within iocage.  Thus I'm carefully not going to be dogmatic about there being a problem.  But I changed something with the patch!  Unfortunately, I can't go back and check, since I've rolled back the jail that I was working in yesterday.

**Comment by nxg** [comment]
I've now had the chance to check this again.  See below.

The correct proxy variables are defined:

    .../site-packages/iocage_lib# export|grep proxy
    http_proxy=http://blah....
    https_proxy=http://blah...

With the distributed ioc_exec.py, the `iocage fetch` appears to work, initially:

    .../site-packages/iocage_lib# iocage fetch
    [0] 11.2-RELEASE
    [1] 11.3-RELEASE
    [2] 11.4-RELEASE
    [3] 12.0-RELEASE
    [4] 12.1-RELEASE
    [5] 12.2-RELEASE
    [6] 12.3-RELEASE
    [7] 13.0-RELEASE
    
    Type the number of the desired RELEASE
    Press [Enter] to fetch the default selection: (13.0-RELEASE)
    Type EXIT to quit: 
    Fetching: 13.0-RELEASE
    
    Downloading: MANIFEST [####################] 100% 
    Downloading: base.txz [####################] 100% 
    Downloading: lib32.txz [####################] 100% 
    Downloading: src.txz [####################] 100% 
    Extracting: base.txz... 
    Extracting: lib32.txz... 
    Extracting: src.txz... 
    
    * Updating 13.0-RELEASE to the latest patch level... 
    Looking up update.FreeBSD.org mirrors... 2 mirrors found.
    Fetching public key from update1.freebsd.org... failed.
    Fetching public key from update2.freebsd.org... failed.
    No mirrors remaining, giving up.
    
    This may be because upgrading from this platform (amd64)
    or release (13.0-RELEASE) is unsupported by tmpw6rahrso. Only
    platforms with Tier 1 support can be upgraded by tmpw6rahrso.
    See https://www.freebsd.org/platforms/index.html for more info.
    
    If unsupported, FreeBSD must be upgraded by source.
    No updates are available to install.
    Run '/tmp/tmpw6rahrso fetch' first.

...but stalls at update time.  So patch `ioc_exec.py`:

    .../site-packages/iocage_lib# ls
    __init__.py		ioc_common.py		ioc_exceptions.py	ioc_fetch.py		ioc_list.py		ioc_upgrade.py
    __pycache__/		ioc_create.py		ioc_exec.py		ioc_fstab.py		ioc_plugin.py		iocage.py
    ioc_check.py		ioc_debug.py		ioc_exec.py-mod		ioc_image.py		ioc_start.py		release.py
    ioc_clean.py		ioc_destroy.py		ioc_exec.py-orig	ioc_json.py		ioc_stop.py		resource.py
    .../site-packages/iocage_lib# diff ioc_exec.py{,-mod}
    79a80,90
    >         # pass on http_proxy or HTTP_PROXY from the environment
    >         # (both are used, but 'curl' prefers lowercase)
    >         su_env.setdefault('http_proxy',
    >                           os.environ.get('http_proxy',
    >                                          os.environ.get('HTTP_PROXY', '')))
    >         su_env.setdefault('https_proxy',
    >                           os.environ.get('https_proxy',
    >                                          os.environ.get('HTTPS_PROXY', '')))
    >         su_env.setdefault('no_proxy',
    >                           os.environ.get('no_proxy',
    >                                          os.environ.get('NO_PROXY', '')))
    .../site-packages/iocage_lib# cp ioc_exec.py{-mod,}  

And try again:

    .../site-packages/iocage_lib# iocage fetch           
    [0] 11.2-RELEASE
    [1] 11.3-RELEASE
    [2] 11.4-RELEASE
    [3] 12.0-RELEASE
    [4] 12.1-RELEASE
    [5] 12.2-RELEASE
    [6] 12.3-RELEASE
    [7] 13.0-RELEASE
    
    Type the number of the desired RELEASE
    Press [Enter] to fetch the default selection: (13.0-RELEASE)
    Type EXIT to quit: 
    Fetching: 13.0-RELEASE
    
    Extracting: base.txz... 
    Extracting: lib32.txz... 
    Extracting: src.txz... 
    
    * Updating 13.0-RELEASE to the latest patch level... 
    Looking up update.FreeBSD.org mirrors... 2 mirrors found.
    Fetching public key from update1.freebsd.org... done.
    Fetching metadata signature for 13.0-RELEASE from update1.freebsd.org... done.
    Fetching metadata index... done.
    Fetching 2 metadata files... done.
    Inspecting system... done.
    Preparing to download files... done.
    Fetching 240 patches.....10....20....30....40....50....60....70....80....90....100....110....120....130....140....150....160....170....180....190....200....210....220....230....240 done.
    Applying patches... done.
    Fetching 15 files... ....10.. done.
    The following files will be removed as part of updating to
    13.0-RELEASE-p6:
    /etc/ssl/certs/080911ac.0
    /etc/ssl/certs/0b7c536a.0
    /etc/ssl/certs/0c4c9b6c.0
    ...
    /usr/src/usr.sbin/bhyve/pci_virtio_9p.c
    /usr/src/usr.sbin/bhyve/pci_virtio_console.c
    /usr/src/usr.sbin/bhyve/pci_virtio_rnd.c
    /usr/src/usr.sbin/bhyve/pci_virtio_scsi.c
    Installing updates...Scanning /zroot/iocage/releases/13.0-RELEASE/root/usr/share/certs/blacklisted for certificates...
    Scanning /zroot/iocage/releases/13.0-RELEASE/root/usr/share/certs/trusted for certificates...
     done.
    .../site-packages/iocage_lib# iocage --version
    Version	1.2
    .../site-packages/iocage_lib# freebsd-version
    13.0-RELEASE-p6
